### PR TITLE
common, os/bluestore: Fix lack of checking for compat on DENC_START

### PR DESCRIPTION
**Pull request by aclamk**
Patch an omission in _denc_start(... decode ...).
There was no checking for v < compat.
The result is that old version will happily attempt to decode messages from outside range.

Affected ceph versions: Luminous to Squid.

Specifically affected types:
```
osd/osd_types.h-  DENC(osd_reqid_t, v, p) {
osd/osd_types.h:    DENC_START(2, 2, p);
```
All other instances are `DENC_START(1, 1, ...)`.
It means that we have no compatibility issues yet, but bumping up compat will cause problem.

Fixes: https://tracker.ceph.com/issues/66157

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/include/denc.h [hunk 1]
```diff
@@ -1818,8 +1818,13 @@ inline std::enable_if_t<traits::supported && !traits::featured> decode_nohead(
 			  __u8 *struct_compat,				\
 			  char **start_pos,				\
 			  uint32_t *struct_len) {			\
+    __u8 code_v = *struct_v;						\
     denc(*struct_v, p);							\
     denc(*struct_compat, p);						\
+    if (code_v < *struct_compat)					\
+      throw ::ceph::buffer::malformed_input(fmt::format(		\
```
Comment (rzarzynski) [inline src/include/denc.h:1825]: IIUC `struct_compat` is a way to let an encoder decide whether a decoder at particular version should refuse-to-decode new bytestream. This is quite fragile as it establishes an interface spanning over the time dimension.

This change would fix only new decoders; in this aspect we don't have any control over the past ones WHILE the interface for current-and-newer encoders still claims we actually do. This can be confusing and dangerous.

In my understanding `DENC` simply doesn't support the `struct_compat` versioning. Shouldn't we be more explicit about that? How about removing the parameter (only **at the interface layer**, bytestreams must be obviously compatible)? Later we could introduce a new macro called e.g. `DENC_WITH_MIN_COMPAT_SINCE_SQUID()`.

**Review by rzarzynski — COMMENTED** [review]
`DENC` for `osd_reqid_t` has been introduced  in 2016 by aa4ab12c64a411a87c4e030e9e280dea19ab23b9. Even before it was at-and-requiring version 2. There is no change in this aspect since many years. However, I'm still worried about the future. Please see the comment below.

File: src/include/denc.h (hunk 1 above)
Comment (aclamk) [reply src/include/denc.h:1825]: The interface for DENC_START always claimed to address forward compatibility problem, but it did not do it.
This already is confusing and dangerous.
Introduction of any new macro that produces backward-compatible bitstream works to visually signal developer, but does not protect against old decoders.
If we are to signal developers, existing `DENC_START` should get renamed to `DENC_START_UNSAFE`.

To make version change reliable we need new encoding that poisons old decoders.
It will be used when bumping up version.
That is, if we transition from `v=1` to `v=2`, the new bitstream will kill old decoder, while the new one should be able to read both.

Comment (rzarzynski) [reply src/include/denc.h:1825]: >  This already is confusing and dangerous.

Currently it's _consistently_ (across all versions) confusing and dangerous.
After the patch it would become more _surprising_ – the check would _usually work_ (especially on developer's desk) but not _always_.
Maybe altering the interface to honor what's possible to implement and what's not is the way to go.

To what extent honor? An encoder could detect (thankfully to feature bits) whether the decoder handles the compat version. The question is whether  it's really useful. I could imagine an interface (only for `min_compat`-taking variant?) that would both visually inform and enforce to handle (or ignore) broken decoder explicitly. However, this might be too much of complexity.

**Comment by aclamk** [comment]
@rzarzynski I added protection against bumping up compat before CEPH_VERSION 21 (U******).
The unfortunate downside is that I had to include "ceph_ver.h" in "denc.h".
The result is that scope of dependencies from "ceph_ver.h" is now much larger.
If needed, this could be improved by splitting commit-related parts of ceph_ver.h and version-related parts of ceph_ver.h

File: src/include/denc.h [hunk 2]
```diff
@@ -1845,11 +1846,26 @@ inline std::enable_if_t<traits::supported && !traits::featured> decode_nohead(
...
+// DENC_START interface suggests it is checking compatibility,
+// but the feature was unimplemented until SQUID.
+// Due to -2 compatibility rule we cannot bump up compat until U____ release.
+
 #define DENC_START(v, compat, p)					\
   __u8 struct_v = v;							\
   __u8 struct_compat = compat;						\
   char *_denc_pchar;							\
   uint32_t _denc_u32;							\
+  static_assert(CEPH_RELEASE >= 21 || compat == 1);			\
```
Comment (rzarzynski) [inline src/include/denc.h:1858]: I think it's a pretty reasonable compromise – although users are made aware about the problem a bit later, at compile time, the code churn is minimal. :+1:

Comment (rzarzynski) [inline src/include/denc.h:1858]: `21` assumes the patch is released with squid and this should be the case.
However, for the sake of contingency planning, let's add comment explaining the math to not forget.

File: src/include/denc.h [hunk 3]
```diff
@@ -51,6 +51,7 @@
 
 #include "common/convenience.h"
 #include "common/error_code.h"
+#include "ceph_ver.h"
```
Comment (rzarzynski) [inline src/include/denc.h:54]: This has been discussed in the last core syncup. `denc.h` is included almost everywhere, and thus `ceph_ver.h` will be too. The problem is we need to rebuild far, far more when it changes. @aclamk is addressing the issue.

**Comment by aclamk** [comment]
Reduced dependency by splitting ceph_ver.h into ceph_ver.h and ceph_release.h 

229 targets dependent on denc.h
```
[akupczyk@o08 build]$ touch ../src/include/denc.h
[akupczyk@o08 build]$ ninja ceph_test_objectstore
[229/229] Linking CXX executable bin/ceph_test_objectstore

```
31 dependent on git SHA1
```
[akupczyk@o08 build]$ git commit --amend
[akupczyk@o08 build]$ ninja ceph_test_objectstore
[0/1] Re-running CMake...
...
[31/31] Linking CXX executable bin/ceph_test_objectstore

```

File: src/test/test_denc.cc [hunk 4]
```diff
@@ -352,23 +352,37 @@ struct foo_t {
...
+    ::denc(v.b, p);
+    if (struct_v >= 2) {
+      ::denc(v.c, p);
+    }
     DENC_FINISH(p);
   }
+};
+WRITE_CLASS_DENC_BOUNDED(foo2_accept1_t)
 
-  friend bool operator==(const foo2_t& l, const foo2_t& r) {
```
Comment (ronen-fr) [inline src/test/test_denc.cc:366]: why was this removed here?

File: src/include/encoding.h [hunk 5]
```diff
@@ -1468,6 +1468,10 @@ decode(std::array<T, N>& v, bufferlist::const_iterator& p)
 #define DECODE_ERR_OLDVERSION(func, v, compatv)					\
   (std::string(func) + " no longer understand old encoding version " #v " < " + std::to_string(compatv))
```
Comment (ronen-fr) [inline src/include/encoding.h:1469]: should be 'understands'

File: src/include/encoding.h [hunk 6]
```diff
@@ -1468,6 +1468,10 @@ decode(std::array<T, N>& v, bufferlist::const_iterator& p)
...
 
+#define DECODE_ERR_NO_COMPAT(func, code_v, v, compatv)					\
```
Comment (ronen-fr) [inline src/include/encoding.h:1471]: pls change cant to cannot

File: src/include/denc.h (hunk 2 above)
Comment (aclamk) [reply src/include/denc.h]: Done? Please take a look.

File: src/include/encoding.h (hunk 5 above)
Comment (aclamk) [reply src/include/encoding.h]: Done.

File: src/include/encoding.h (hunk 6 above)
Comment (aclamk) [reply src/include/encoding.h:1471]: Done, and in denc.h too.

**Comment by aclamk** [comment]
jenkins test make check

**Comment by aclamk** [comment]
jenkins test make check

**Comment by aclamk** [comment]
jenkins test make check

**Comment by aclamk** [comment]
jenkins test api

**Comment by aclamk** [comment]
jenkins test windows

**Comment by aclamk** [comment]
jenkins test make check

**Comment by aclamk** [comment]
jenkins test make check

**Comment by aclamk** [comment]
@rzarzynski This PR passed rados suite. 
https://pulpito.ceph.com/akupczyk-2024-07-12_08:52:40-rados-aclamk-testing-ganymede-2024-07-10-1144-distro-default-smithi/
https://pulpito.ceph.com/akupczyk-2024-07-11_11:24:14-rados-aclamk-testing-ganymede-2024-07-10-1144-distro-default-smithi/

Do we want to merge it now?

File: src/test/test_denc.cc [hunk 7]
```diff
@@ -741,3 +755,42 @@ TEST(denc, no_copy_if_segmented_and_lengthy)
...
+  bufferlist bl;
+  {
+    auto app = bl.get_contiguous_appender(s);
+    denc(v2, app);
+  }
+
+  foo_t v1;
+  v1.a = 111; v1.b = 111;
+  auto bpi = bl.front().begin();
+  ASSERT_ANY_THROW(denc(v1,bpi));
```
Comment (rzarzynski) [inline src/test/test_denc.cc:795]: ACK.

File: src/osd/osd_types.h [hunk 8]
```diff
@@ -164,7 +164,7 @@ struct osd_reqid_t {
   {}
 
   DENC(osd_reqid_t, v, p) {
-    DENC_START(2, 2, p);
+    DENC_START_COMPAT_2(2, 2, p);
```
Comment (rzarzynski) [inline src/osd/osd_types.h:167]: I'm afraid `osd_reqid_t` is exposed also to clients (see `osdc`), so we cannot count on the `n+2` rule in this particular case.

File: src/ceph_ver.h.in.cmake [hunk 9]
```diff
@@ -3,8 +3,7 @@
 
 #define CEPH_GIT_VER @CEPH_GIT_VER@
 #define CEPH_GIT_NICE_VER "@CEPH_GIT_NICE_VER@"
-#define CEPH_RELEASE @CEPH_RELEASE@
-#define CEPH_RELEASE_NAME "@CEPH_RELEASE_NAME@"
-#define CEPH_RELEASE_TYPE "@CEPH_RELEASE_TYPE@"
+
+#include "ceph_release.h"
```
Comment (rzarzynski) [inline src/ceph_ver.h.in.cmake:7]: `doc/dev/release-checklists.rst` is probably affected. I would be fine with a followup PR if needed.

File: src/include/denc.h [hunk 10]
```diff
@@ -1865,7 +1866,7 @@ inline std::enable_if_t<traits::supported && !traits::featured> decode_nohead(
   __u8 struct_compat = compat;						\
   char *_denc_pchar;							\
   uint32_t _denc_u32;							\
-  static_assert(CEPH_RELEASE >= 21 || compat == 2);			\
+  static_assert(CEPH_RELEASE >= (19/*squid*/ + 2) || compat == 2);	\
```
Comment (rzarzynski) [inline src/include/denc.h:1875]: `common/ceph_releases.h` has the helpers:

```cpp
// the C++ version of CEPH_RELEASE_* defined by include/rados.h
enum class ceph_release_t : std::uint8_t {
  unknown = 0,
  argonaut,
  bobtail,
  cuttlefish,
  dumpling,
  emperor,
  firefly,
  giant,
  hammer,
  infernalis,
  jewel,
  kraken,
  luminous,
  mimic,
  nautilus,
  octopus,
  pacific,
  quincy,
  reef,
  squid,
  max,
};
```

Anyway, this comment is less-than-nit.

**Review by rzarzynski — COMMENTED** [review]
From the comments in this review I perceive the `osd_reqid_t`-related as the only important one.

After solving, let's freeze the PR and, if possible, merge it atomically with the `squid` backport.

File: src/include/denc.h [hunk 11]
```diff
@@ -1862,17 +1863,28 @@ inline std::enable_if_t<traits::supported && !traits::featured> decode_nohead(
...
+  do {
+
+// For osd_reqid_t which cannot be upgraded at all.
+// We used it to communicate with clients and now we cannot safely upgrade.
+#define DENC_START_OSD_REQID(v, compat, p)				\
+  __u8 struct_v = v;							\
+  __u8 struct_compat = compat;						\
+  char *_denc_pchar;							\
+  uint32_t _denc_u32;							\
+  static_assert(compat == 2, "osd_reqid_t cannot be upgraded");		\
```
Comment (rzarzynski) [inline src/include/denc.h:1887]: This is very pessimistic stance as it assumes never-ending compatibility for clients. There are multiple different views on the that but I'm OK with being most cautious here – we can change it later if needed.

**Review by rzarzynski — APPROVED** [review]

**Comment by aclamk** [comment]
https://pulpito.ceph.com/akupczyk-2024-07-17_11:23:18-rados-wip-aclamk-denc-compat-check-squid-distro-default-smithi/
https://pulpito.ceph.com/akupczyk-2024-07-17_09:03:38-rados-wip-aclamk-denc-compat-check-squid-distro-default-smithi/
1 failure that seems unrelated.
